Complete yarn 3 upgrade for projects (#3879)
plone/volto: docs/source/upgrade-guide/index.md

2. Change your root project `Makefile` to include these commands:

> ```diff

Reviewer: This diff needs to be indented so that it is part of the numbered list.

Hint: I always indent my list items to 4 spaces, so that when I need to indent code blocks, I can select the selection and hit tab to indent it properly.

Author: All indentations fixed. What do you mean by "I always indent my list items to 4 spaces"? the whole 1. xxxx element?  the markdown linter complains if I do so. or just the text?

Reviewer: Like so, where on the first line, `Item 1` is indented 4 spaces, counting the `1`, `.`, and two spaces:

````md
1.  Item 1
1.  Item 2

    a.  Item 2.a.

        ```shell
        rm -rf  # twitter lulz
        ```

1.  Item 3
````

Some code editors get annoying with their auto-correct or auto-formatting when you mix the number of spaces between levels of indentation. The editor sometimes inconsistently forces 2, 3, or 4 spaces on indents, or a mixture of them, especially on new lines or select, then tab to indent.

+.PHONY: install
+install: ## Install the frontend
+       @echo "Install frontend"
+       $(MAKE) omelette
+       $(MAKE) preinstall

> +       $(MAKE) develop

Reviewer: Is it possible that a developer would want to run `make omelette`, `make preinstall`, or `make develop` directly? If so, then there should be help text for each target below.

Author: ups... I've found a bug in the middle. Thanks to point me to double check this. Commit incoming!

+       yarn install
+
+.PHONY: preinstall
+preinstall:
+       if [ -f $$(pwd)/mrs.developer.json ]; then if [ -f $$(pwd)/node_modules/.bin/missdev ]; then yarn develop; else yarn develop:npx; fi; fi
+
+.PHONY: develop
+develop:
+       npx -p mrs-developer missdev --config=jsconfig.json --output=addons --fetch-https
+
+.PHONY: omelette
+omelette:
+       if [ ! -d omelette ]; then ln -sf node_modules/@plone/volto omelette; fi
+
+.PHONY: patches
+patches:
+       /bin/bash patches/patchit.sh > /dev/null 2>&1 ||true
```

You can copy the file over to your project if you have not made any amendment to it.

3. Change your `package.json` scripts section:

> ```diff

Reviewer: Indent this diff code block.

Author: Done.

   "version": "1.0.0",
   "scripts": {
     "start": "razzle start",
-    "preinstall": "if [ -f $(pwd)/mrs.developer.json ]; then if [ -f $(pwd)/node_modules/.bin/missdev ]; then yarn develop; else yarn develop:npx; fi; fi",
-    "omelette": "if [ ! -d omelette ]; then ln -sf node_modules/@plone/volto omelette; fi",
-    "patches": "/bin/bash patches/patchit.sh > /dev/null 2>&1 ||true",
-    "postinstall": "yarn omelette && yarn patches",
+    "postinstall": "make omelette && make patches",
-    "patches": "/bin/bash patches/patchit.sh > /dev/null 2>&1 ||true",
```

> Yarn 3 no longer support inline bash scripts in the `scripts` section.

Reviewer: ```suggestion
Yarn 3 no longer supports inline bash scripts in the `scripts` section.
```

It does not support the `preinstall` lifecycle state, so we cannot trigger things while we run `yarn install` or just `yarn`.

> We need to move all the commands related to `preinstall` actions to `Makefile` and update the calls in both files as shown above.

Reviewer: ```suggestion
As a result, we moved all the commands related to the `preinstall` actions into `Makefile`, then updated the calls in both files as shown above.
```

```{important}

> After making the changes in this step, we will have to modify our development workflow, specially if we are making use of actions that happened during the `preinstall` state, the most relevant one being updating the code `mrs-developer` is managing.

Reviewer: ```suggestion
After making the changes in this step, we will have to modify our development workflow, especially if we use actions that happened during the `preinstall` state.
The most relevant one is updating the code that is managed by `mrs-developer`.
```
